File: ascii_fields/animations/doppler.py
import math

from ..core import Animation, BLACK_BG, RESET, clamp, density_char, gray_fg, resolve_theme


class _Front:
  __slots__ = ("x", "y", "vx", "vy", "t0")

  def __init__(self, x, y, vx, vy, t0):
    self.x = x
    self.y = y
    self.vx = vx
    self.vy = vy
    self.t0 = t0


def _shift_color(shift):
  # shift: 0 = redshift, 0.5 = none, 1 = blueshift
  if shift < 0.5:
    t = shift / 0.5
    return (255, int(70 + 185 * t), int(70 + 185 * t))
  t = (shift - 0.5) / 0.5
  return (int(255 - 215 * t), int(255 - 150 * t), 255)


class DopplerAnimation(Animation):
  """The radial-velocity wobble of a star tugged by an orbiting exoplanet. The
  star circles the shared barycentre and its light ripples outward; the side it
  moves toward is blueshifted, the side it recedes from is redshifted. With a
  truecolor terminal the shift is shown in colour."""

  thresholds = [
    (0.10, " "), (0.22, "."), (0.34, ":"), (0.46, "-"), (0.58, "="),
    (0.70, "+"), (0.82, "*"), (0.92, "#"), (1.01, "@"),
  ]

  C = 0.42                  # wave speed (normalised units / s)
  EMIT_DT = 0.14

  def __init__(self):
    self._fronts = []
    self._next = 0.0
    self._last = 0.0

  def render(self, width, height, elapsed, phase, options):
    if elapsed < self._last:
      self._fronts = []
      self._next = 0.0
    self._last = elapsed

    ax = width / max(1, height * 2.0)
    omega = 1.4
    r_star = 0.18
    r_planet = 0.62
    # star wobble and (opposite-phase) planet
    sx = r_star * math.cos(omega * elapsed)
    sy = r_star * math.sin(omega * elapsed)
    svx = -r_star * omega * math.sin(omega * elapsed)
    svy = r_star * omega * math.cos(omega * elapsed)
    px = -r_planet * math.cos(omega * elapsed)
    py = -r_planet * math.sin(omega * elapsed)

    while elapsed >= self._next:
      tt = self._next
      ssx = r_star * math.cos(omega * tt)
      ssy = r_star * math.sin(omega * tt)
      vvx = -r_star * omega * math.sin(omega * tt)
      vvy = r_star * omega * math.cos(omega * tt)
      self._fronts.append(_Front(ssx, ssy, vvx, vvy, tt))
      self._next += self.EMIT_DT
    self._fronts = [f for f in self._fronts if elapsed - f.t0 < 5.0]

    mono = resolve_theme(options, "scene") == "mono"
    bright = options.brightness
    contrast = options.contrast
    rows = []
    for row in range(height):
      cy = (row / max(1, height - 1) - 0.5) * 2.0
      line = [BLACK_BG]
      last = None
      for col in range(width):
        cx = (col / max(1, width - 1) - 0.5) * 2.0 * ax
        intensity = 0.0
        dop_acc = 0.0
        for f in self._fronts:
          radius = self.C * (elapsed - f.t0)
          dx = cx - f.x
          dy = cy - f.y
          dist = math.hypot(dx, dy) or 1e-4
          b = math.exp(-((dist - radius) ** 2) / 0.0010)
          if b > 0.001:
            intensity += b
            dop = (f.vx * dx + f.vy * dy) / (dist * self.C)
            dop_acc += b * dop
        # star and planet markers
        intensity += 1.0 * math.exp(-(((cx - sx) ** 2 + (cy - sy) ** 2)) / 0.0016)
        intensity += 0.5 * math.exp(-(((cx - px) ** 2 + (cy - py) ** 2)) / 0.0010)
        value = clamp(intensity * contrast)
        ch = density_char(value, self.thresholds)
        if value <= 0.001:
          line.append(" ")
          last = None
          continue
        if mono:
          color = gray_fg(value, 234, 255, bright)
          if color != last:
            line.append(f"\x1b[38;5;{color}m")
            last = color
        else:
          shift = clamp(0.5 + 0.5 * (dop_acc / intensity if intensity else 0.0) * 1.6)
          r, g, b = _shift_color(shift)
          rgb = (int(r * value), int(g * value), int(b * value))
          if rgb != last:
            line.append(f"\x1b[38;2;{rgb[0]};{rgb[1]};{rgb[2]}m")
            last = rgb
        line.append(ch)
      line.append(RESET)
      rows.append("".join(line))
    return "\n".join(rows)

File: src/animations/doppler.rs
//! Exoplanet Doppler: a star wobbles, wavefronts bunch on the approach side
//! (blueshift) and stretch behind (redshift). Custom truecolor renderer so the
//! shift sign maps to red/blue regardless of theme; falls back to gray when
//! the theme is `grayscale`/`mono`.

use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, density_char, gray_fg, resolve_theme, BLACK_BG, RESET};
use std::fmt::Write;

const TH: &[(f64, char)] = &[
  (0.10, ' '),
  (0.22, '.'),
  (0.34, ':'),
  (0.46, '-'),
  (0.58, '='),
  (0.70, '+'),
  (0.82, '*'),
  (0.92, '#'),
  (1.01, '@'),
];
const C: f64 = 0.42;
const EMIT_DT: f64 = 0.14;

#[derive(Clone, Copy)]
struct Front {
  x: f64,
  y: f64,
  vx: f64,
  vy: f64,
  t0: f64,
}

pub struct Doppler {
  fronts: Vec<Front>,
  next_emit: f64,
  last: f64,
}
impl Default for Doppler {
  fn default() -> Self {
    Self { fronts: Vec::new(), next_emit: 0.0, last: 0.0 }
  }
}

fn shift_color(shift: f64) -> (u8, u8, u8) {
  if shift < 0.5 {
    let t = shift / 0.5;
    (255, (70.0 + 185.0 * t) as u8, (70.0 + 185.0 * t) as u8)
  } else {
    let t = (shift - 0.5) / 0.5;
    ((255.0 - 215.0 * t) as u8, (255.0 - 150.0 * t) as u8, 255)
  }
}

impl Animation for Doppler {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    if ctx.elapsed < self.last {
      self.fronts.clear();
      self.next_emit = 0.0;
    }
    self.last = ctx.elapsed;
    let w = ctx.width;
    let h = ctx.height;
    let ax = w as f64 / (h as f64 * 2.0).max(1.0);
    let omega = 1.4_f64;
    let r_star = 0.18;
    let r_planet = 0.62;
    let sx = r_star * (omega * ctx.elapsed).cos();
    let sy = r_star * (omega * ctx.elapsed).sin();
    let px_pl = -r_planet * (omega * ctx.elapsed).cos();
    let py_pl = -r_planet * (omega * ctx.elapsed).sin();
    while ctx.elapsed >= self.next_emit {
      let tt = self.next_emit;
      let ssx = r_star * (omega * tt).cos();
      let ssy = r_star * (omega * tt).sin();
      let vvx = -r_star * omega * (omega * tt).sin();
      let vvy = r_star * omega * (omega * tt).cos();
      self.fronts.push(Front { x: ssx, y: ssy, vx: vvx, vy: vvy, t0: tt });
      self.next_emit += EMIT_DT;
    }
    self.fronts.retain(|f| ctx.elapsed - f.t0 < 5.0);
    let theme = resolve_theme(&ctx.options.theme, "scene");
    let mono = theme == "mono";
    let bright = ctx.options.brightness;
    let contrast = ctx.options.contrast;
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    for row in 0..h {
      out.push_str(BLACK_BG);
      let cy = (row as f64 / dh - 0.5) * 2.0;
      let mut last_color: Option<u8> = None;
      let mut last_rgb: Option<(u8, u8, u8)> = None;
      for col in 0..w {
        let cx = (col as f64 / dw - 0.5) * 2.0 * ax;
        let mut intensity = 0.0_f64;
        let mut dop_acc = 0.0_f64;
        for f in &self.fronts {
          let radius = C * (ctx.elapsed - f.t0);
          let dx = cx - f.x;
          let dy = cy - f.y;
          let dist = (dx * dx + dy * dy).sqrt().max(1e-4);
          let b = (-((dist - radius).powi(2)) / 0.0010).exp();
          if b > 0.001 {
            intensity += b;
            let dop = (f.vx * dx + f.vy * dy) / (dist * C);
            dop_acc += b * dop;
          }
        }
        intensity += (-((cx - sx).powi(2) + (cy - sy).powi(2)) / 0.0016).exp();
        intensity += 0.5 * (-((cx - px_pl).powi(2) + (cy - py_pl).powi(2)) / 0.0010).exp();
        let value = clamp(intensity * contrast);
        let ch = density_char(value, TH);
        if value <= 0.001 {
          out.push(' ');
          last_color = None;
          last_rgb = None;
          continue;
        }
        if mono {
          let color = gray_fg(value, 234, 255, bright);
          if Some(color) != last_color {
            let _ = write!(out, "\x1b[38;5;{}m", color);
            last_color = Some(color);
          }
        } else {
          let shift = clamp(0.5 + 0.5 * (if intensity > 0.0 { dop_acc / intensity } else { 0.0 }) * 1.6);
          let (r, g, b) = shift_color(shift);
          let rgb = ((r as f64 * value) as u8, (g as f64 * value) as u8, (b as f64 * value) as u8);
          if Some(rgb) != last_rgb {
            let _ = write!(out, "\x1b[38;2;{};{};{}m", rgb.0, rgb.1, rgb.2);
            last_rgb = Some(rgb);
          }
        }
        out.push(ch);
      }
      out.push_str(RESET);
      if row + 1 < h {
        out.push('\n');
      }
    }
  }
}
